src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/DecryptorPalOpenSsl.Decrypt.cs

```diff
@@ -0,0 +1,124 @@
+                        // the information by parsing it and to store the appropriate keys to call the EVP_PKEY_derive method on them
+                        // to obtain the secret and proceed with manual decryption. For further documentation see:
+                        // https://wiki.openssl.org/index.php/EVP_Key_Agreement
+                        exception = new PlatformNotSupportedException();
+                        break;
+
+                    default:
+                        // As we commented in the Windows implementation:
+                        // Since only the framework can construct RecipientInfo's, we're at fault if we get here. So it's okay to assert and throw rather than
+                        // returning to the caller.
+                        Debug.Fail($"Unexpected RecipientInfoType: {type}");
+                        throw new NotSupportedException();
+                }
+            }
+
+            if (exception != null)
+                return null;
+
+            return decryptedContent;
+        }
+
+        private ContentInfo TryDecryptTrans(X509Certificate2 recipientCert, out Exception exception)
+        {
+            // As of now, KeyTransport recipients have RSA as their only choice for the symmetric key encryption,
+            // but to future proof, might as well check so that we don't try to get a private RSA key from something that doesn't have one.
+            if (recipientCert.GetKeyAlgorithm() != Oids.Rsa)
+            {
+                exception = new PlatformNotSupportedException();
+                return null;
+            }
+
+            RSAOpenSsl kekAlgoPKey = recipientCert.GetRSAPrivateKey() as RSAOpenSsl;
+            byte[] decryptedMessage = null;
+
+            using (SafeBioHandle decryptedBuffered = Interop.Crypto.CreateMemoryBio())
+            using (SafeX509Handle recipientCertHandle = Interop.Crypto.X509Duplicate(recipientCert.Handle))
+            using (SafeEvpPKeyHandle pKeyKandle = kekAlgoPKey.DuplicateKeyHandle())
+            {
+                Interop.Crypto.CheckValidOpenSslHandle(decryptedBuffered);
+                Interop.Crypto.CheckValidOpenSslHandle(recipientCertHandle);
+                Interop.Crypto.CheckValidOpenSslHandle(pKeyKandle);
+
+                // If there's no content OpenSSL will fail to decrypt, so do the check manually before delegating.
+                ContentInfo currentContent = _decodedMessage.GetEmbeddedContent();
+                if (currentContent.Content.Length == 0)
+                {
+                    exception = null;
+                    return currentContent;
+                }
+                
+                int status = Interop.Crypto.CmsDecrypt(_decodedMessage, recipientCertHandle, pKeyKandle, decryptedBuffered);
+
+                exception = status == 1 ?
+                    null :
+                    new CryptographicException(
```

Comment:
Windows gets the HR for the last message and puts it in a CryptographicException. I fel like the CryptographicException was important so I maintained it, as for the message, checking the messages windows might return in that case I felt that there were so many reasons the message would fail that maybe giving the OpenSSL message would be clearer.